rfd/0139-gogo-must-go.md

```diff
@@ -0,0 +1,132 @@
+library and subsequently ran into issues during `gogo/protobuf`'s deprecated. `csproto`
+is a library that elects to use the proper protobuf runtime for the appropriate message.
+As a result, you could have an intermingling of different protobuf libraries without compatibility
+issues.
+
+### Our complications
+
+Teleport has a number of complications tied in with our use of `gogo/protobuf` that will make
+our migration unique. Along with our complications will be listed a number of potential
+mitigation strategies to discuss.
+
+**The primary issue here is our use of `gogo/protobuf` extensions.** We should strive
+to remove these extensions while maintaining backwards compatibility.
+
+#### JSON/YAML (de)serialization (removing `gogoproto.jsontag`)
+
+We rely heavily on `gogo/protobuf`'s custom JSON tags for serialization our messages into JSON.
+Additionally, we use a library called [`ghodss/yaml`](https://github.com/ghodss/yaml) to create YAML,
+which also relies on these JSON tags.
+
+Each resource is serialized into JSON before being pushed into our backend, which means the
+protobuf JSON tags have an impact quite far down our stack. There are a number of potential
+mitigations here with differing levels of impact. The primary goal of any mitigation strategy
+here should be **migrate away from gogoproto.jsontag extensions.**
+
+##### Serialize objects in backend as the gRPC wire format
+
+Serializing the objects in the backend as the gRPC wire format has a number of benefits:
+
+- Seems to remove the need for most unmarshaling functions in `lib/services`. (confirm?)
+- The backend has no need for JSON marshaling/unmarshaling, so JSON changes will not
+  affect the backend.
+- Allows us to remove gogoproto extensions without affecting backend storage.
+- Allows us to utilize fields in the backend more efficiently and store larger objects.
+
+The primary disadvantage here is that manually examining the backend is more difficult as
+the stored values will no longer be human readable.
+
+##### Write custom marshalers for objects
+
+In order to keep our user facing representations of the objects, we will need to write
+custom JSON and YAML marshalers for our objects. This may be time intensive but not overall
+a significant amount of work.
+
+#### Casting types (removing `gogoproto.casttype`)
+
+We additionally rely heavily on `gogoproto.casttype`, which allows us to avoid doing explicit
+casts or conversions within go code. We will have to take on this conversion work, unfortunately.
+
+#### Other gogoproto tags
+
+##### `gogoproto.nullable`
+
+We will need to remove the `gogoproto.nullable` tags, which overall shouldn't be a huge impact.
+We should migrate to using `optional` as defined [here](https://protobuf.dev/programming-guides/proto3/#field-labels).
+
+##### `gogoproto.stdtime`
+
+We will need to migrate to using `timestamppb` directly instead of relying on gogo's `stdtime` tag.
```

Comment:
Is this migration something we can start doing now ? Would this be a wire-compatible change ?

Reply:
AFAIK all of these changes should be wire compatible. This is something that we can do now (I'm putting together a POC for it as we speak, actually).

Reply:
You're not allowed to break protobuf wire compatibility anyway. 😎